[Spring 지하철 노선도 - 1,2단계] 스컬(김윤기) 미션 제출합니다.

skullkim opened the pull request:
앨런! 이번 미션 잘부탁해요ㅎㅎ

HJ-Rich committed:
feat(station): Station equals&hashCode 재정의

HJ-Rich committed:
feat(controller-advice): RestControllerAdvice 구현

HJ-Rich committed:
feat(exception): 지하철역 생성 시 중복에 대한 예외 추가

HJ-Rich committed:
feat(station-dao): 지하철역 추가시 이름 중복 검증 로직 추가

HJ-Rich committed:
feat(stations): 역들을 관리하는 Stations 추가, 지하철역 추가 기능 구현

HJ-Rich committed:
feat(no-station-found-exception): 커스텀 예외 추가

찾고자 하는 지하철역이 없을 때 던져지는 커스텀 예외 추가

HJ-Rich committed:
feat(stations): 지하철역 조회, 삭제 기능 구현

HJ-Rich committed:
feat(station): 아이디 일치여부 확인 메서드 추가

HJ-Rich committed:
feat(station-dao): 지하철역 삭제, 조회 기능 및 테스트 구현

HJ-Rich committed:
feat(exception): 요청 자원이 없을 경우에 대한 커스텀 예외 추가

HJ-Rich committed:
feat(line-dao): 노선을 관리하는 LineDao 구현

HJ-Rich committed:
refactor(line-response, line-request): 노선 요청, 응답 Dto 생성자 추가

HJ-Rich committed:
feat(controller-advice): 맵핑 예외 추가

HJ-Rich committed:
feat(station-controller): 단일 지하철역 삭제 구현

HJ-Rich committed:
feat(line): 노선을 의미하는 Line 도메인 객체 구현

HJ-Rich committed:
feat(line-controller): 노선을 관리하는 LineController 구현

HJ-Rich committed:
feat(properties): 프로파일, schema.sql 추가

HJ-Rich committed:
refactor(spring): Dao 에 대한 Spring Framework 적용

HJ-Rich committed:
refactor(stations): 사용되지 않는 도메인 객체 제거

HJ-Rich committed:
fix(application.yml): 테스트 DB 설정 오류 수정

HJ-Rich committed:
feat(station): StationService 구현

HJ-Rich committed:
feat(line): LineService 구현

hongbin-dev commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,75 @@
+@Repository
+public class LineDao {
+
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    private final SimpleJdbcInsert simpleInsert;
+    private final JdbcTemplate jdbcTemplate;
```
NamedParameterJdbcTemplate가 있는데, jdbcTemplate를 사용하시는 이유가 있을까요?
jdbcTemplate는 제외해도 좋을 것 같습니다!

skullkim replied:
수정하겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/dao/LineDao.java:
```diff
@@ -0,0 +1,75 @@
+    public void deleteById(Long id) {
+        final String sql = "delete from LINE where id = ?";
+        jdbcTemplate.update(sql, id);
+    }
```
식별자에 맞는 로우가 없어도, 성공하는 메서드일 것 같은데요. 의도하신 걸까요?

skullkim replied:
아니요.. 맞는 로우가 없으면 예외를 던지는 방식으로 바꾸겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,48 @@
+        return name;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof Line)) {
+            return false;
+        }
+
+        Line line = (Line) o;
+        return name.equals(line.name);
+    }
+
+    @Override
+    public int hashCode() {
+        return name != null ? name.hashCode() : 0;
+    }
```
name을 식별자로 둔 이유가 있을까요?

skullkim replied:
디비에 역을 저장하기 전까지는 id를 가지지 않고 name은 고유하기 때문에 좀 더 범용성 있는 name을 식별자로 사용하게 되었습니다.

hongbin-dev replied:
객체 내에 ID가 있는이상, 식별자를 id로 하는게 좋을 것 같습니다.
영속성레이어와 애플리케이션의 식별자의 불일치해서 '의도하는대로 동작안할 수도 있겠다' 라는 생각이 들었어요

skullkim replied:
수정하겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Station.java:
```diff
@@ -1,5 +1,7 @@
 package wooteco.subway.domain;
 
+import java.util.Objects;
+
```
Station의 생성자가 중복이 많네요. 부생성자, 주생성자를 this()를 통해서 묶을 수 있을 것 같아요.

skullkim replied:
수정하겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/dto/LineRequest.java:
```diff
@@ -18,6 +18,10 @@ public LineRequest(String name, String color, Long upStationId, Long downStation
         this.distance = distance;
     }
 
+    public LineRequest(String name, String color) {
+        this(name, color, null, null, 0);
+    }
```
`null, null, 0`를 쓰신 이유가 있을까요?
사용하지 않는 필드라면 제거하는게 좋을 것 같아요.

skullkim replied:
미션 처음 시작 부터 있었던 코드이고, 당장은 사용하지 않지만 이후의 지하철 미션에서 사용될 것으로 보입니다. 미션 시작 때 주어진 코드를 맘대로 삭제해도 되나요?

hongbin-dev replied:
네 변경해도 되는걸로 알고있습니다!
베이스 코드였군요 👀

skullkim replied:
수정하겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/dto/LineResponse.java:
```diff
@@ -10,13 +9,20 @@ public class LineResponse {
     private String color;
     private List<StationResponse> stations;
 
+    public LineResponse() {
+    }
```
기본생성자가 필요한 이유가 있을까요?

skullkim replied:
LineResponse를 반환값으로, 정확히는 JSON응답에 사용하기 위해 추가했습니다.
@RestController를 사용하는 클래스에서 다음과 같은 코드가 있다면
```java
@GetMapping(value = "/lines/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<LineResponse> showLine(@PathVariable Long id) {
        Line line = lineService.findById(id);
        return ResponseEntity.ok().body(getLineResponse(line));
    }
```
ReponseEntity는 내부적으로 Jackson을 사용해 객체를 JSON으로 변환하는 것으로 알고 있습니다. 이 때, Jackson의 올바른 동작을 위해 기본 생성자가 필요해서 사용하게 되었습니다. 기본생성자를 사용하지 않는다면 다른 생성자에 @JsonCreator 어노테이션을 붙여 사용되는 되는데 혹이 이것을 사용하라는 의도인가요?

hongbin-dev replied:
아 아닙니다! Response객체는 컨트롤러에서 직렬화해서 응답할 때, 기본생성자없이도 가능해서요.

인수테스트에서 역직렬화를 하고 있어서 문제가 발생하는군요? 그럼 유지하셔도 될 것 같습니다.

hongbin-dev commented on src/main/java/wooteco/subway/dto/LineResponse.java:
```diff
@@ -10,13 +9,20 @@ public class LineResponse {
         this.stations = stations;
     }
 
+    public LineResponse(Long id, String name, String color) {
+        this(id, name, color, new ArrayList<>());
+    }
```
마이너한 이슈지만 매번 `new ArrayList<>()` 생성하지 않고, `List.of()`로 만들어진 객체를 사용할 수 있습니다.

skullkim replied:
수정하겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/exception/LineDuplicateException.java:
```diff
@@ -0,0 +1,10 @@
+
+    private static final String DUPLICATED_MESSAGE = "이미 존재하는 노선입니다.";
+
+    public LineDuplicateException() {
+        super(DUPLICATED_MESSAGE);
+    }
```
에러로그가 조금 부실한거 같아요!
어떤 값 때문에 문제가 발생하는지 수정하면, 빠르게 피드백받을 수 있을 것 같아요.

skullkim replied:
수정하겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -0,0 +1,18 @@
+import wooteco.subway.domain.Line;
+import wooteco.subway.dto.LineRequest;
+
+public interface LineService {
```
인터페이스로 분리하신 이유가 뭘까요?

skullkim replied:
서비스와 컨트롤러 계층을 분리하기 위해 인터페이스를 사용했습니다. 서비스 구현체를 그대로 사용한다면 컨트롤러가 서비스에 의존하게 되서 서비스의 변경으로 인해 컨트롤러까지 변경을 해야 한다고 생각했습니다. 따라서 인터페이스를 사용해 컨트롤러와 서비스 계층을 분리해보았습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/SpringLineService.java:
```diff
@@ -0,0 +1,43 @@
+import wooteco.subway.dto.LineRequest;
+
+@Service
+public class SpringLineService implements LineService {
```
클래스 이름이 모호한거 같아요. 서비스의 구현체는 프레임워크마다 생성되는걸까요?

hongbin-dev commented on src/main/java/wooteco/subway/service/SpringLineService.java:
```diff
@@ -0,0 +1,43 @@
+import wooteco.subway.domain.Line;
+import wooteco.subway.dto.LineRequest;
+
+@Service
+public class SpringLineService implements LineService {
```
LineService도 트랜잭션을 달아볼 수 있겠네요!

skullkim replied:
찾아보니 h2는 mysql의 스토리엔진은 InnoDB로 설정했을 때와 달리 자동으로 트랜잭션 롤백을 지원해주지 않네요. 수정하겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/SpringLineService.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public Line save(LineRequest lineRequest) {
+        Line line = new Line(null, lineRequest.getName(), lineRequest.getColor());
```
```suggestion
        Line line = new Line(lineRequest.getName(), lineRequest.getColor());
```
도메인의 편의생성자를 추가해보는건 어떨까요?

skullkim replied:
이 방식이 더 좋겠네요. 수정하겠습니다,

hongbin-dev commented on src/main/java/wooteco/subway/service/SpringLineService.java:
```diff
@@ -0,0 +1,43 @@
+        return lineDao.findAll();
+    }
+
+    @Override
+    public void update(Long id, LineRequest lineRequest) {
+        lineDao.update(id, lineRequest);
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        lineDao.deleteById(id);
+    }
```
식별자에 해당하는 row가 없으면 어떻게되나요?

skullkim replied:
식별자에 해당하는 row가 없으면 예외를 던지는 방식으로 바꾸겠습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/SpringLineService.java:
```diff
@@ -0,0 +1,43 @@
+        this.lineDao = lineDao;
+    }
+
+    @Override
+    public Line save(LineRequest lineRequest) {
+        Line line = new Line(null, lineRequest.getName(), lineRequest.getColor());
+        return lineDao.save(line);
+    }
```
애플리케이션에서 중복제어를 하지 않는 이유가 있을까요?

skullkim replied:
필요 없을거 같아서 안했습니다. 디비, 프레임워크를 모두 제외하고 순수 자바 객체로만 이루어진 도메인 로직만 생각 했을 때에는 애플리케이션에서 중복 제어를 하는 것이 맞다고 생각 합니다. 하지만 모든 데이터를 디비에 저장하고 있는 상황에서 애플리케이션에서 중복을 제어 한다면 디비에 있는 데이터를 도메인 객체에 할당하고 중복 검사를 해야 합니다, 그 후 중복이 발생했다면 예외를 던지고 그에 맞는 처리를 해야죠.  디비에서 컬럼을 unique으로 하고 단순히 insert를 해도 똑같이 예외가 던져지고 그에 대한 처리를 할 수 있습니다. 그러면 굳이 데이터를 디비에서 꺼내고, 도메인 객체에 할당하고, 중복 검사를 하는 로직을 수행할 이유가 있을까요?

도메인이 애플리케이션에서 핵심 로직을 구현해야 한다는 것에는 동의합니다. 하지만 디비가 애플리케이션에 추가됨으로 인해서 도메인과 디비 사이에서 애매한 로직들이 생성되는 것 같습니다. 그 케이스 중 하나가 save() 메서드와 같이 입력한 데이터를 기존 데이터와 단순 비교하는 거라 생각합니다. 이럴때는 코드가 단순해 지는 방향을 선택하는 것이 좋다고 생각합니다. 도메인이 간단해야 파악하기 쉬우니까요. 

앨런의 생각은 어떤가요? 앨런의 생각도 듣고 싶어요.

hongbin-dev replied:
전 개인적으로 추가했으면 하는 편입니다! 몇가지 이유가 있는데요.

'비즈니스 규칙은 애플리케이션에 들어가야한다'라고 생각하는데요. 만약 처음입사한 개발자가 코드만보고 비즈니스를 모두 파악할 수 있어야한다고 생각합니다.
프로젝트에 처음 참여하는 개발자라면, 'Line이 중복제어는 없구나?'라는 생각이 들 것 같아요. 

데이터베이스에 의존적인 코드가 될 것 같아요. 만약 시스템이 커져서 데이터베이스를 NoSQL로 변경하게되면 어떻게될까요?  NoSQL에서 제공하는 중복제어기능을 사용해야할까요?

한번 중복을 조회하고, 생성하는게 큰 비용은 아닌 것 같아요. 또 중복제어를 Service레이어에서 진행할텐데, 애매한 코드가 발생할까요..? 이것조차 도메인코드라고 생각하는 편입니다.

```
public void save(...) {
  validateName(name)
  stations.save(new Line(name))
}
```

hongbin-dev commented on src/main/java/wooteco/subway/ui/StationControllerAdvice.java:
```diff
@@ -0,0 +1,21 @@
+@RestControllerAdvice
+public class StationControllerAdvice {
+
+    @ExceptionHandler({StationDuplicateException.class, LineDuplicateException.class, NoLineFoundException.class,
+            NoStationFoundException.class, DuplicateKeyException.class})
+    public ResponseEntity<Void> duplicateStation() {
+        return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+    }
```
다른 RuntimeException, Exception 객체들도 핸들링해보면 좋을 것 같아요.

skullkim replied:
그래야 의도치 않은 에러도 핸들할 수 있겠네요. 수정하겠습니다.

hongbin-dev commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,219 @@
+import org.springframework.http.MediaType;
+import wooteco.subway.dto.LineResponse;
+
+@DisplayName("노선 관련 기능")
+public class LineAcceptanceTest extends AcceptanceTest {
```
update, delete시 식별자가 존재하지 않으면 발생하는 예외도 잡아볼 수 있을 것 같아요.

skullkim replied:
수정하겠습니다.

hongbin-dev commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,219 @@
+import org.springframework.http.MediaType;
+import wooteco.subway.dto.LineResponse;
+
+@DisplayName("노선 관련 기능")
+public class LineAcceptanceTest extends AcceptanceTest {
```
행위가 많아지다보니 개인적으로 가독성이 떨어지는 것 같아요.
각각의 행위를 메소드로 분리하거나, 다이나믹테스트를 적용해봐도 좋을 것 같아요.

https://tecoble.techcourse.co.kr/post/2020-07-31-dynamic-test/

skullkim replied:
이런것도 있네요, 수정하겠습니다.

hongbin-dev commented on src/test/java/wooteco/subway/dao/LineDaoTest.java:
```diff
@@ -0,0 +1,116 @@
+        lineDao = new LineDao(namedParameterJdbcTemplate, dataSource, jdbcTemplate);
+    }
+
+    @Nested
+    @DisplayName("새로운 노선을 저장할 때")
+    class SaveTest {
```
`@Nested` 👍

hongbin-dev commented on src/test/resources/schema.sql:
```diff
@@ -0,0 +1,24 @@
+create table if not exists STATION
+(
+    id   bigint auto_increment not null,
+    name varchar(255)          not null unique,
+    primary key (id)
+);
+
+create table if not exists LINE
+(
+    id    bigint auto_increment not null,
+    name  varchar(255)          not null unique,
+    color varchar(20)           not null,
+    primary key (id)
+);
+
+create table if not exists SECTION
+(
+    id              bigint auto_increment not null,
+    line_id         bigint                not null,
+    up_station_id   bigint                not null,
+    down_station_id bigint                not null,
+    distance        int,
+    primary key (id)
+);
```
기본설정이 production의 resources/schema.sql 를 바라보고 있어서 제거해도 좋을 것 같아요.

skullkim replied:
테스트를 위해 별도의 shcema를 설정하는 것이 좋지 않을까요? 지금이야 production과 test 둘 다 in-memory h2를 사용하지만 production은 원칙적으로는 mysql같은 디비를 사용해야 한다고 생각합니다. 테스트용 schema를 설정하기 않는다면 production용 디비가 바뀌었을 때 테스트가 깨지게 되겠죠. 코드상에 변경이 발생한게 아닌 디비의 변경으로 인해 테스트가 깨져도 되는 건가요?

hongbin-dev replied:
몇 가지 솔루션이 있는데요!

- h2에서 mysql 문법을 호환하게 사용할 수 있습니다.
- 회사마다 다르겠지만 운영에서 Mysql를 사용한다면, 영속성을 사용하는 통합테스트는 Mysql을 띄워서 최대한 동일한 환경에서 테스트합니다.

그래서 이 부분은 걱정할 포인트가 아니라고 생각했어요! 개인적으로 운영/테스트 스키마를 한벌로 관리했으면 좋겠습니다! (두벌로 관리하면 번거로워서..)

제가 피드백 남긴 건, 해당 `schema.sql`이 테스트에 적용이 안되는점이에요. 해당 파일을 삭제/주석/변경해도 테스트가 잘 되더라고요 👀

hongbin-dev reviewed (changes requested):
안녕하세요 스컬!
1, 2 단계 미션 잘 구현해주셨군요 💯💯
피드백과 질문 남겼으니 확인해주세요!!

skullkim committed:
refactor(LineDao.java): 필요 없는 필드 삭제

다른 필드로 대체 가능한 필드 삭제

BREAKING CHANGE: JdbcTemplate 필드 삭제
as-is:
@Repository
public class LineDao {

    ...
    private final JdbcTemplate jdbcTemplate;

    public LineDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate, DataSource dataSource,
                   JdbcTemplate jdbcTemplate) {
        ...
        this.jdbcTemplate = jdbcTemplate;
    }

    ...

     public void deleteById(Long id) {
        final String sql = "delete from LINE where id = ?";
        jdbcTemplate.update(sql, id);
    }

    public void deleteAll() {
        final String sql = "delete from LINE";
        jdbcTemplate.update(sql);
    }
}

to-be:
@Repository
public class LineDao {

    ...

    public LineDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate, DataSource dataSource) {
        ...
    }

    ...
}

skullkim committed:
refactor(Station.java): 생성자 수정

생성자를 부생성자와 주생성자로 분리

BREKAING CHANGE: 기존 3개 생성자 중 두 개를 부생성자로, 하나를
주생성자로 사용

as-is:
public class Station {
    private Long id;
    private String name;

    public Station() {
    }

    public Station(Long id, String name) {
        this.id = id;
        this.name = name;
    }

    public Station(String name) {
        this.name = name;
    }

    ...
}

to-be:
public class Station {

    private static final  long EMPTY_ID = 0;
    private static final String EMPTY_NAME = "";

    private final Long id;
    private final String name;

    public Station() {
        this(EMPTY_ID, EMPTY_NAME);
    }

    public Station(Long id, String name) {
        this.id = id;
        this.name = name;
    }

    public Station(String name) {
        this(EMPTY_ID, name);
    }

    ...
}

skullkim committed:
refactor(LineResponse.java): 생성자 수정

부 생성자에서 빈 리스트를 매번 생성하는 대신 기존에 생성된 빈 리스트를
재사용 하는 방식으로 수정

BREAKING CHANGE: 새로운 빈 리스트를 생성하는 방식에서 List.of()를
사용하는 방식으로 수정

as-is:
public class LineResponse {
    ...

    public LineResponse(Long id, String name, String color) {
        this(id, name, color, new ArrayList<>());
    }

    ...
}

to-be:
public class LineResponse {
    ...

     public LineResponse(Long id, String name, String color) {
        this(id, name, color, List.of());
    }

    ...
}

skullkim committed:
refactor(Line.java): 부생성자 생성

서비스에서 Line 인스턴스 생성을 편하게 하기 위해 부생성자 생성

BREAKING CHANGE: id를 인자로 받지 않는 부생성자 생성
as-is:
public class Line {

    ...

    public Line(Long id, String name, String color) {
        this.id = id;
        this.name = name;
        this.color = color;
    }

    ...

}

@Service
public class SpringLineService implements LineService {
    ...

    @Override
    public Line save(LineRequest lineRequest) {
        Line line = new Line(null, lineRequest.getName(), lineRequest.getColor());
        ...
    }

    ...

}

to-be:
public class Line {
    ...

    public Line(Long id, String name, String color) {
        this.id = id;
        this.name = name;
        this.color = color;
    }

    public Line(String name, String color) {
        this(null, name, color);
    }

    ...
}

@Service
public class SpringLineService implements LineService {
    ...

    @Override
    public Line save(LineRequest lineRequest) {
        Line line = new Line(lineRequest.getName(), lineRequest.getColor());
        ...
    }

    ...
}

skullkim committed:
refactor(SpringLineService.java): 트랜잭션 추가

트랜잭션 롤백을 적용하기 위해 @Transactional 어노테이션 추가

BREAKING CHANGE: 디비를 접근하는 메서드에 @Transactional 어노테이션 추가
as-is:
@Service
public class SpringLineService implements LineService {

    ...

    @Override
    public Line save(LineRequest lineRequest) {
        ...
    }

    @Override
    public Line findById(Long id) {
        ...
    }

    @Override
    public List<Line> findAll() {
        ...
    }

    @Override
    public void update(Long id, LineRequest lineRequest) {
        ...
    }

    @Override
    public void deleteById(Long id) {
        ...
    }
}

to-be:
@Service
public class SpringLineService implements LineService {

    ...

    @Override
    @Transactional
    public Line save(LineRequest lineRequest) {
        ...
    }

    @Override
    @Transactional(readOnly = true)
    public Line findById(Long id) {
        ...
    }

    @Override
    @Transactional(readOnly = true)
    public List<Line> findAll() {
        ...
    }

    @Override
    @Transactional
    public void update(Long id, LineRequest lineRequest) {
        ...
    }

    @Override
    @Transactional
    public void deleteById(Long id) {
        ...
    }
}

skullkim committed:
feat(LineDao.java): 예외 추가

update와 delete 시 영향을 받은 레코드가 존재하지 않는다면 예외를 던지는
기능 추가

skullkim committed:
refactor(LineDao.java): 라인저장에 대한 예외 변경

라인 저장 시 예외가 발생하면 예외가 발생한 데이터를 메시지에 추가하기
위한 로직 추가

BREAKING CHANGE: save() 메서드에 try-catch 로직 추가
as-is:
@Repository
public class LineDao {
    ...

    public Line save(Line line) {
        final Map<String, Object> params = new HashMap<>();
        params.put("name", line.getName());
        params.put("color", line.getColor());
        final Long id = simpleInsert.executeAndReturnKey(params).longValue();
        return new Line(id, line.getName(), line.getColor());
    }

    ...
}

to-be:
@Repository
public class LineDao {
    ...

    public Line save(Line line) {
        ...
        Long id;
        try {
            id = simpleInsert.executeAndReturnKey(params).longValue();
        } catch (DuplicateKeyException e) {
            throw new LineDuplicateException(line.toString());
        }
        return new Line(id, line.getName(), line.getColor());
    }

    ...
}

public class Line {
    ...

     @Override
    public String toString() {
        return "Line{" +
                "name='" + name + '\'' +
                ", color='" + color + '\'' +
                '}';
    }
}

skullkim committed:
refactor(SpringLineServiceTest.java): 테스트 추가

dao에서 예외를 던지는 로직 추가로 인한 테스트 추가

BREAKING CHANGE: 라인 업데이트, 삭제, 저장에 대한 예외 케이스 테스트
추가
as-is:
@SpringBootTest
@Transactional
class SpringLineServiceTest {
    ...

    @Nested
    @DisplayName("새로운 노선을 저장할 때")
    class SaveTest {

        ...

        @Test
        @DisplayName("노선 이름이 중복되면 예외가 발생한다.")
        void save_Fail_If_Exists() {
            ...
            assertThatThrownBy(() -> lineService.save(LINE_FIXTURE))
                    .isInstanceOf(DuplicateKeyException.class);
        }
    }

    ...

    @Test
    @DisplayName("아이디로 지하철노선을 삭제할 수 있다")
    void deleteById() {
        ...
    }

    @Test
    @DisplayName("노선 이름과 색상을 변경할 수 있다")
    void update() {
       ...
    }
}

to-be:
@SpringBootTest
@Transactional
class SpringLineServiceTest {
    ...

    @Nested
    @DisplayName("새로운 노선을 저장할 때")
    class SaveTest {

        ...

        @Test
        @DisplayName("노선 이름이 중복되면 예외가 발생한다.")
        void save_Fail_If_Exists() {
            ...
            assertThatThrownBy(() -> lineService.save(LINE_FIXTURE))
                    .isInstanceOf(LineDuplicateException.class)
                    .hasMessageContaining("이미 존재하는 노선입니다. Line{name='2호선', color='bg-color-600'}");
        }
    }

    ...

    @Nested
    @DisplayName("아이디로 지하철노선을 삭제할 때")
    class DeleteLineTest {

        @Test
        @DisplayName("아이디가 존재하면 아이디로 지하철노선을 삭제할 수 있다.")
        void deleteById() {
            ...
        }

        @Test
        @DisplayName("아이디가 존재하지 않는다면 예외를 던진다.")
        void delete_By_Id_Fail() {
            assertThatThrownBy(() -> lineService.deleteById(1L))
                    .isInstanceOf(NoLineFoundException.class)
                    .hasMessageContaining("요청한 노선이 존재하지 않습니다. id=1");
        }
    }

    @Nested
    @DisplayName("노선 이름과 색상을 변경하려할 때")
    class UpdateLineTest {

        @Test
        @DisplayName("노선이 존재하면 노선 이름과 색상을 변경할 수 있다")
        void update_Line_Success() {
            ...
        }

        @Test
        @DisplayName("노선이 존재하지 않으면 예외를 던진다.")
        void update_Line_Fail() {
            assertThatThrownBy(() -> lineService.update(1L, new LineRequest("a", "b")))
                    .isInstanceOf(NoLineFoundException.class)
                    .hasMessageContaining("요청한 노선이 존재하지 않습니다. id=1 LineRequest{name='a', color='b'}");

        }
    }
}

skullkim committed:
refactor(StationDao.java): 필요 없는 필드 삭제

다른 필드로 대체 가능한 필드 삭제

BREAKING CHANGE: JdbcInseret 필드 삭제
as-is:
@Repository
public class StationDao {
    ...
    private final JdbcTemplate jdbcTemplate;

    public StationDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate, DataSource dataSource,
                      JdbcTemplate jdbcTemplate) {
        ...
        this.jdbcTemplate = jdbcTemplate;
    }

    ...

    public void deleteById(Long id) {
        ...
        jdbcTemplate.update(sql, id);
    }
}

to-be:
@Repository
public class StationDao {
    ...

    public StationDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate, DataSource dataSource) {
        ...
    }

    public void deleteById(Long id) {
        ...
        namedParameterJdbcTemplate.update(sql, Map.of("id", id));
    }
}

skullkim committed:
feat(StationServiceTest.java): 테스트 추가

dao에서 예외를 던지는 로직 추가고 인한 테스트 추가

skullkim committed:
feat(StationDao.java): 예외 추가

핸들링이 필요한 예외 추가

BREAKING CHANGE: 역 삭제와 중복된 역 저장에 대한 예외 추가

skullkim committed:
fix(Station.java): toString 추가

예외 메시지에서 사용될 Station 인스턴스 상태를 위해 toString 재정의

BREAKING CHANGE: toString 재정의
public class Station {
    ...

    @Override
    public String toString() {
        return "Station{" +
                "name='" + name + '\'' +
                '}';
    }
}

skullkim committed:
refactor(dao/): 코드 간소화

지하철 라인과 역을 저장하를 로직 간소화

BREAKING CHANGE: 필요 없는 로직 삭제
as-is:
@Repository
public class StationDao {
    ...

    public Station save(Station station) {
        ...
        final Long id;
        try {
            id = simpleInsert.executeAndReturnKey(params).longValue();
        } catch (DuplicateKeyException e) {
            throw new StationDuplicateException(station.toString());
        }
        return new Station(id, station.getName());
    }

    ...
}

@Repository
public class LineDao {
    ...

    public Line save(Line line) {
        ...
        Long id;
        try {
            id = simpleInsert.executeAndReturnKey(params).longValue();
        } catch (DuplicateKeyException e) {
            throw new LineDuplicateException(line.toString());
        }
        return new Line(id, line.getName(), line.getColor());
    }

    ...
}

to-be:
@Repository
public class StationDao {
    ...

    public Station save(Station station) {
        ...
        try {
            final Long id = simpleInsert.executeAndReturnKey(params).longValue();
            return new Station(id, station.getName());
        } catch (DuplicateKeyException e) {
            throw new StationDuplicateException(station.toString());
        }
    }

    ...
}

@Repository
public class LineDao {
    ...

    public Line save(Line line) {
        ...
        try {
            final Long id = simpleInsert.executeAndReturnKey(params).longValue();
            return new Line(id, line.getName(), line.getColor());
        } catch (DuplicateKeyException e) {
            throw new LineDuplicateException(line.toString());
        }
    }

    ...
}

skullkim committed:
feat(StationControllerAdvice): 에러 응답 수정

에러 원인을 바디에 추가, 예상치 못한 exception에 대한 handler 추가

skullkim committed:
refactor(LineAcceptanceTest.java): 중복 로직 제거

각 테스트에서 라인을 생성하는 로직을 하나의 메서드로 통합

BREAKING CAHNGE: 라인을 생성하는 로직을 addLine()메서드로 통합
as-is:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {

    @DisplayName("노선을 생성한다.")
    @Test
    void createLine() {
        // given
        Map<String, String> params = new HashMap<>();
        params.put("name", "신분당선");
        params.put("color", "bg-red-600");

        // when
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();

        ...
    }

    @DisplayName("기존에 존재하는 노선 이름으로 노선 생성 시도 시 Bad request가 응답된다.")
    @Test
    void createLineWithDuplicateName() {
        // given
        Map<String, String> params = new HashMap<>();
        params.put("name", "강남역");
        params.put("color", "bg-red-600");
        RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();

        // when
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then()
                .log().all()
                .extract();

        ...
    }

    @DisplayName("노선을 조회한다.")
    @Test
    void getLines() {
        /// given
        Map<String, String> params1 = new HashMap<>();
        params1.put("name", "1호선");
        params1.put("color", "bg-red-600");
        ExtractableResponse<Response> createResponse1 = RestAssured.given().log().all()
                .body(params1)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();

        Map<String, String> params2 = new HashMap<>();
        params2.put("name", "2호선");
        params2.put("color", "bg-green-600");
        ExtractableResponse<Response> createResponse2 = RestAssured.given().log().all()
                .body(params2)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();

        ...
    }

    @DisplayName("노선을 제거한다.")
    @Test
    void deleteLine() {
        // given
        Map<String, String> params = new HashMap<>();
        params.put("name", "1호선");
        params.put("color", "bg-red-600");
        ExtractableResponse<Response> createResponse = RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();

        ...
    }

    @DisplayName("특정 노선을 조회한다.")
    @Test
    void getLine() {
        /// given
        Map<String, String> params1 = new HashMap<>();
        params1.put("name", "1호선");
        params1.put("color", "bg-red-600");
        ExtractableResponse<Response> createResponse = RestAssured.given().log().all()
                .body(params1)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();

        ...
    }

     @DisplayName("노선을 갱신한다.")
    @Test
    void updateLine() {
        // given
        Map<String, String> params = new HashMap<>();
        params.put("name", "1호선");
        params.put("color", "bg-red-600");
        ExtractableResponse<Response> createResponse = RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();

        ...
    }
}

to-be:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {

    @DisplayName("노선을 생성한다.")
    @Test
    void createLine() {
        ExtractableResponse<Response> response = addLine("신분당선", "bg-red-600");

        ...
    }

    @DisplayName("기존에 존재하는 노선 이름으로 노선 생성 시도 시 Bad request가 응답된다.")
    @Test
    void createLineWithDuplicateName() {
        addLine("강남역", "bg-red-600");
        ExtractableResponse<Response> response = addLine("강남역", "bg-red-600");

        ...
    }

     @DisplayName("노선을 조회한다.")
    @Test
    void getLines() {
        ExtractableResponse<Response> createResponse1 = addLine("1호선", "bg-red-600");
        ExtractableResponse<Response> createResponse2 = addLine("2호선", "bg-green-600");
        ...
    }

      @DisplayName("노선을 제거한다.")
    @Test
    void deleteLine() {
        ExtractableResponse<Response> createResponse = addLine("1호선", "bg-red-600");
        ...
    }

     @DisplayName("특정 노선을 조회한다.")
    @Test
    void getLine() {
        ExtractableResponse<Response> createResponse = addLine("1호선", "bg-red-600");
        ...
    }

     @DisplayName("노선을 갱신한다.")
    @Test
    void updateLine() {
        ExtractableResponse<Response> createResponse = addLine("1호선", "bg-red-600");
        ...
    }

    private ExtractableResponse<Response> addLine(final String name, final String color) {
        Map<String, String> params = new HashMap<>();
        params.put("name", name);
        params.put("color", color);
        return RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();
    }
}

skullkim committed:
refactor(LineAcceptanceTest.java): 코드 가독성 개선

중복 로직을 메서드로 묶어 코드 가독성 개선

BREAKING CHANGE: getLines() 메서드, updateLine() 메서드 추가
as-is:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {

    ...

    @DisplayName("노선을 조회한다.")
    @Test
    void getLines() {
        ...

        // when
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .when()
                .get("/lines")
                .then().log().all()
                .extract();

        // then
        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
        List<Long> expectedLineIds = Arrays.asList(createResponse1, createResponse2).stream()
                .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
                .collect(Collectors.toList());
        List<Long> resultLineIds = response.jsonPath().getList(".", LineResponse.class).stream()
                .map(it -> it.getId())
                .collect(Collectors.toList());
        assertThat(resultLineIds).containsAll(expectedLineIds);
    }

    ...

     @DisplayName("특정 노선을 조회한다.")
    @Test
    void getLine() {
        ...

        // when
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .when()
                .get(createResponse.header("Location"))
                .then().log().all()
                .extract();
        final LineResponse actual = createResponse.jsonPath().getObject(".", LineResponse.class);

        ...
    }

    @DisplayName("노선을 갱신한다.")
    @Test
    void updateLine() {
        ...

        // when
        Map<String, String> newParams = new HashMap<>();
        final String newLineName = "11호선";
        newParams.put("name", newLineName);
        newParams.put("color", "bg-red-600");
        String uri = createResponse.header("Location");
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .body(newParams)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .put(uri)
                .then().log().all()
                .extract();

        ExtractableResponse<Response> updatedResponse = RestAssured.given().log().all()
                .when()
                .get(uri)
                .then().log().all()
                .extract();
        final LineResponse actual = updatedResponse.jsonPath().getObject(".", LineResponse.class);

        ...
    }

    ...
}

to-be:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {
    ...

    @DisplayName("노선을 조회한다.")
    @TestFactory
    Stream<DynamicTest> getLines() {
        ExtractableResponse<Response> createResponse1 = addLine("1호선", "bg-red-600");
        ExtractableResponse<Response> createResponse2 = addLine("2호선", "bg-green-600");

        return Stream.of(
                DynamicTest.dynamicTest("생성된 노선 목록을 불러온다", () -> {
                    ExtractableResponse<Response> response = getLines("/lines");

                    assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
                }),
                DynamicTest.dynamicTest("생성된 노선이 저장한 노선과 일치한지 확인한다.", () -> {
                    ExtractableResponse<Response> response = getLines("/lines");
                    List<Long> expectedLineIds = Stream.of(createResponse1, createResponse2)
                            .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
                            .collect(Collectors.toList());
                    List<Long> resultLineIds = response.jsonPath().getList(".", LineResponse.class).stream()
                            .map(it -> it.getId())
                            .collect(Collectors.toList());

                    assertThat(resultLineIds).containsAll(expectedLineIds);
                })
        );
    }

    @DisplayName("특정 노선을 조회한다.")
    @Test
    void getLine() {
        ...

        ExtractableResponse<Response> response = getLines(createResponse.header("Location"));
        final LineResponse actual = response.jsonPath().getObject(".", LineResponse.class);

        ...
    }

    @DisplayName("노선을 갱신한다.")
    @Test
    void updateLine() {
        ...

        final String newLineName = "11호선";
        String uri = createResponse.header("Location");
        ExtractableResponse<Response> response = updateLine(newLineName, "bg-red-600", uri);

        ExtractableResponse<Response> updatedResponse = getLines(uri);
        final LineResponse actual = updatedResponse.jsonPath().getObject(".", LineResponse.class);

        ...
    }

    ...

    private ExtractableResponse<Response> updateLine(final String name, final String color, final String uri) {
        Map<String, String> newParams = new HashMap<>();
        newParams.put("name", name);
        newParams.put("color", "bg-red-600");
        return RestAssured.given().log().all()
                .body(newParams)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .put(uri)
                .then().log().all()
                .extract();

    }

    private ExtractableResponse<Response> getLines(final String url) {
        return RestAssured.given().log().all()
                .when()
                .get(url)
                .then().log().all()
                .extract();
    }
}

skullkim committed:
refactor(StationAcceptanceTest.java): 중복 제거

코드 가독성을 위해 중복 로직을 별도의 메서드로 분리

BREAKING CHANGE: 중복 로직 제거를 위해 addStation(), getStations()
메서드 추가

as-is:
@DisplayName("지하철역 관련 기능")
public class StationAcceptanceTest extends AcceptanceTest {

    @DisplayName("지하철역을 생성한다.")
    @Test
    void createStation() {
        // given
        Map<String, String> params = new HashMap<>();
        params.put("name", "강남역");

        // when
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();

        ...
    }

    @DisplayName("기존에 존재하는 지하철역 이름으로 지하철역을 생성한다.")
    @Test
    void createStationWithDuplicateName() {
        // given
        Map<String, String> params = new HashMap<>();
        params.put("name", "강남역");
        RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();

        // when
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then()
                .log().all()
                .extract();

        ...
    }

    @DisplayName("지하철역을 조회한다.")
    @Test
    void getStations() {
        /// given
        Map<String, String> params1 = new HashMap<>();
        params1.put("name", "강남역");
        ExtractableResponse<Response> createResponse1 = RestAssured.given().log().all()
                .body(params1)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();

        Map<String, String> params2 = new HashMap<>();
        params2.put("name", "역삼역");
        ExtractableResponse<Response> createResponse2 = RestAssured.given().log().all()
                .body(params2)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();

        // when
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .when()
                .get("/stations")
                .then().log().all()
                .extract();

        // then
        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
        List<Long> expectedLineIds = Arrays.asList(createResponse1, createResponse2).stream()
                .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
                .collect(Collectors.toList());
        List<Long> resultLineIds = response.jsonPath().getList(".", StationResponse.class).stream()
                .map(StationResponse::getId)
                .collect(Collectors.toList());
        assertThat(resultLineIds).containsAll(expectedLineIds);
    }

    @DisplayName("지하철역을 제거한다.")
    @Test
    void deleteStation() {
        // given
        Map<String, String> params = new HashMap<>();
        params.put("name", "강남역");
        ExtractableResponse<Response> createResponse = RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();

        // when
        String uri = createResponse.header("Location");
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .when()
                .delete(uri)
                .then().log().all()
                .extract();

        ...
    }
}

to-be:
@DisplayName("지하철역 관련 기능")
public class StationAcceptanceTest extends AcceptanceTest {

    @DisplayName("지하철역을 생성한다.")
    @Test
    void createStation() {
        ExtractableResponse<Response> response = addStation("강남역");

        ...
    }

    @DisplayName("기존에 존재하는 지하철역 이름으로 지하철역을 생성한다.")
    @Test
    void createStationWithDuplicateName() {
        addStation("강남역");
        ExtractableResponse<Response> response = addStation("강남역");

        ...
    }

    @DisplayName("지하철역을 조회한다.")
    @Test
    Stream<DynamicTest> getStations() {
        ExtractableResponse<Response> createResponse1 = addStation("강남역");
        ExtractableResponse<Response> createResponse2 = addStation("역삼역");

        return Stream.of(
                DynamicTest.dynamicTest("생성된 지하철역 목록을 불러온다.", () -> {
                    ExtractableResponse<Response> response = getStations("/stations");
                    assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
                }),
                DynamicTest.dynamicTest("생성된 지하철역 목록이 저장한 목록과 일치한지 확인한다.", () -> {
                    ExtractableResponse<Response> response = getStations("/stations");
                    List<Long> expectedLineIds = Arrays.asList(createResponse1, createResponse2).stream()
                            .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
                            .collect(Collectors.toList());
                    List<Long> resultLineIds = response.jsonPath().getList(".", StationResponse.class).stream()
                            .map(StationResponse::getId)
                            .collect(Collectors.toList());
                    assertThat(resultLineIds).containsAll(expectedLineIds);
                })
        );
    }

    @DisplayName("지하철역을 제거한다.")
    @Test
    void deleteStation() {
        ExtractableResponse<Response> createResponse = addStation("강남역");

        String uri = createResponse.header("Location");
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .when()
                .delete(uri)
                .then().log().all()
                .extract();

        ...
    }

    private ExtractableResponse<Response> addStation(final String name) {
        Map<String, String> params = new HashMap<>();
        params.put("name", name);
        return RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();
    }

    private ExtractableResponse<Response> getStations(final String url) {
        return RestAssured.given().log().all()
                .when()
                .get("/stations")
                .then().log().all()
                .extract();
    }
}

skullkim committed:
chore(src/): final 추가

파라미터, 지역 변수에 final 추가

skullkim committed:
refactor(exception/): 필요 없는 생성자 제거

사용하지 않는 기본 생성자 제거

skullkim committed:
refactor(service/): 인터페이스와 구현체 이름 변경

프레임워크에 의존적인 이름을 사용하지 않기 위헤 인터페이스과 구현체 이름
변경

BREAKING CHANGE: 인터페이스, 구현체 이름  변경
as-is:
interface:
    LineService
    StationService
concrete class:
    SpringLineServic
    SpringStationService

to-be:
interface:
    CommonLineService
    CommonStationService
concrete class:
    LineServic
    StationService

skullkim committed:
refactor(Line.java): 식별자 변경

객체를 식별하는데 사용되는 식별자 변경

BREAKING CHANGE: 식별자를 name에서 id로 변경
as-is:
public class Line {
    ...

    @Override
    public boolean equals(Object o) {
       ...

        Line line = (Line) o;
        return name.equals(line.name);
    }

    @Override
    public int hashCode() {
        return name != null ? name.hashCode() : 0;
    }

    ...
}

to-be:
public class Line {
    ...

    @Override
    public boolean equals(Object o) {
        ...

        Line line = (Line) o;
        return id.equals(line.id);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    ...
}

skullkim committed:
refactor(LineRequest.java): 필요 없는 로직 삭제

당장은 사용하지 않는 필드와 메서드 삭제

BREAKING CHANGE: 상,하행 종점, 거리에 대한 정보를 나타내는 필드와
getter삭제
as-is
public class LineRequest {
    ...
    private Long upStationId;
    private Long downStationId;
    private int distance;

    ...

    public LineRequest(final String name, final String color, final Long upStationId, final Long downStationId,
                       final int distance) {
        this.name = name;
        this.color = color;
        this.upStationId = upStationId;
        this.downStationId = downStationId;
        this.distance = distance;
    }

    public LineRequest(final String name, final String color) {
        this(name, color, null, null, 0);
    }

    public Long getUpStationId() {
        return upStationId;
    }

    public Long getDownStationId() {
        return downStationId;
    }

    public int getDistance() {
        return distance;
    }

    ...
}

to-be:
public class LineRequest {
    ...
}

skullkim committed:
refactor(LineDao.java): 커스텀 예외 삭제

라인에 대한 커스텀 예외 삭제

BREAKING CAHNGE: 라인에 대한 작업시 발생하는 예외를 커스텀 예외에서 표준
예외로 변경
as-is:
@Repository
public class LineDao {
    ...

    public Line save(final Line line) {
        ...
        try {
            ...
        } catch (DuplicateKeyException e) {
            throw new LineDuplicateException(line.toString());
        }
    }

    ...

    public void update(final Long id, final LineRequest lineRequest) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new NoLineFoundException("id=" + id + " " + lineRequest);
        }
    }

    public void deleteById(final Long id) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new NoLineFoundException("id=" + id);
        }
    }
}

to-be:
@Repository
public class LineDao {

    ...
    private static final String LINE_DUPLICATED = "이미 존재하는 노선입니다. ";
    private static final String LINE_NOT_FOUND = "요청한 노선이 존재하지 않습니다. ";

    ...

    public Line save(final Line line) {
        ...
        try {
            ...
        } catch (DuplicateKeyException e) {
            throw new IllegalStateException(LINE_DUPLICATED + line);
        }
    }

    ...

    public void update(final Long id, final LineRequest lineRequest) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(LINE_NOT_FOUND + "id=" + id + " " + lineRequest);
        }
    }

    public void deleteById(final Long id) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(LINE_NOT_FOUND + "id=" + id);
        }
    }
}

skullkim committed:
refactor(StationDao.java): 커스텀예외를 표준 예외로 변경

역에 대한 연산시 발생하는 커스텀 예외를 유지보수성을 위해 표준 예외로
변경

BREAKING CHANGE: 커스텀 예외들을 IllegalStateException으로 변경
as-is:
@Repository
public class StationDao {
    ...

    public Station save(final Station station) {
        ...
        try {
            ...
        } catch (DuplicateKeyException e) {
            throw new StationDuplicateException(station.toString());
        }
    }

    public void deleteById(final Long id) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new NoStationFoundException("id=" + id);
        }
    }
}

to-be:
@Repository
public class StationDao {

    ...
    private static final String STATION_DUPLICATED = "이미 존재하는 지하철역입니다. ";
    private static final String STATION_NOT_FOUND = "요청한 지하철 역이 존재하지 않습니다. ";

    ...

    public Station save(final Station station) {
        ...
        try {
            ...
        } catch (DuplicateKeyException e) {
            throw new IllegalStateException(STATION_DUPLICATED + station);
        }
    }

    ...

    public void deleteById(final Long id) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(STATION_NOT_FOUND + "id=" + id);
        }
    }
}

skullkim committed:
chore(StationControllerAdvice.java): 사용하지 않는 import 삭제

skullkim committed:
refactor(Station.java): 식별자 변경

식별자를 이름에서 id로 변경

BREAKING CHANGE: 식별자를 id로 변경
as-is:
public class Station {
    ...

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Station)) {
            return false;
        }
        Station station = (Station) o;
        return Objects.equals(name, station.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    ...
}

to-be:
public class Station {
    ...

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Station)) {
            return false;
        }
        Station station = (Station) o;
        return id.equals(station.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    ...
}

skullkim committed:
refactor(test/java/wooteco/subway/acceptance): 메서드 이름 변경

의미가 명확하지 않은 메서드 이름 변경

BREAKING CHANGE: 역과 라인을 생성하는 메서드 이름에 사용되는 'add'를
'create'로 변경

as-is:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {
    ...

     private ExtractableResponse<Response> addLine(final String name, final String color) {
        ...
    }

    ...
}

@DisplayName("지하철역 관련 기능")
public class StationAcceptanceTest extends AcceptanceTest {
    ...

    private ExtractableResponse<Response> addStation(final String name) {
        ...
    }

    ...
}

to-be:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {
    ...

    private ExtractableResponse<Response> createLine(final String name, final String color) {
        ...
    }

    ...
}

@DisplayName("지하철역 관련 기능")
public class StationAcceptanceTest extends AcceptanceTest {
    ...

    private ExtractableResponse<Response> createStation(final String name) {
        ...
    }

    ...
}

skullkim committed:
refactor(StaitonAcceptanceTest.java): 메서드 시그니처 수정

모든 지하철 역을 가져오는 메서드 파라미터 삭제

BREAKING CHANGE: getStations() 메서드 파라미터 삭제
as-is:
@DisplayName("지하철역 관련 기능")
public class StationAcceptanceTest extends AcceptanceTest {
    ...

    private ExtractableResponse<Response> getStations(final String url) {
        ...
    }
}

to-be:
@DisplayName("지하철역 관련 기능")
public class StationAcceptanceTest extends AcceptanceTest {
    ...

    private ExtractableResponse<Response> getStations() {
        ...
    }
}

skullkim committed:
refactor(LineAcceptanceTest.java): 메서드 분리

모든 역을 가져오는 역할과 역 하나를 가져오는 역할, 두 개를 하는 하나의
메서드를 두 개의 메서드로 분리

BREAKING CHANGE: getLines()를 두 개의 매서드로 분리
as-is:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {
    ...

     private ExtractableResponse<Response> getLines(final String url) {
        ...
    }
}

to-be:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {
    ...

    private ExtractableResponse<Response> getLines() {
        return RestAssured.given().log().all()
                .when()
                .get("/lines")
                .then().log().all()
                .extract();
    }

    private ExtractableResponse<Response> getLine(final String url) {
        return RestAssured.given().log().all()
                .when()
                .get(url)
                .then().log().all()
                .extract();
    }
}

skullkim committed:
refactor(StationControllerAdvice.java): 클래스 이름 변경

해당 클래스는 모든 컨트롤러에 대한 에러 핸들링을 담당하고 있으므로
특정 컨트롤러를 명시하는 부분을 클래스 이름에서 제거

BREAKING CHANGE: 클래스 이름에서 Station 제거
as-is:
@RestControllerAdvice
public class StationControllerAdvice {
    ...
}

to-be:
@RestControllerAdvice
public class ControllerAdvice {
    ...
}

skullkim committed:
refactor(ui/): 필요없는 어노테이션 속성 삭제

@RestController는 JSON만 받으므로 불필요한 어노테이션 속성 삭제

BREAKING CHANGE: produces=MediaType.APPLICATION_JSON_VALUE 삭제
as-is:
@RestController
public class LineController {
    ...

    @GetMapping(value = "/lines", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<List<LineResponse>> showLines() {
        ...
    }

    @GetMapping(value = "/lines/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<LineResponse> showLine(@PathVariable final Long id) {
        ...
    }

    ...
}

@RestController
public class StationController {
    ...

    @GetMapping(value = "/stations", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<List<StationResponse>> showStations() {
        ...
    }

    ...
}

to-be:
@RestController
public class LineController {
    ...

    @GetMapping(value = "/lines")
    public ResponseEntity<List<LineResponse>> showLines() {
        ...
    }

    @GetMapping(value = "/lines/{id}")
    public ResponseEntity<LineResponse> showLine(@PathVariable final Long id) {
        ...
    }
}

@RestController
public class StationController {
    ...

    @GetMapping(value = "/stations")
    public ResponseEntity<List<StationResponse>> showStations() {
    }

    ...
}

hongbin-dev commented on src/main/java/wooteco/subway/domain/Station.java:
```diff
@@ -3,10 +3,15 @@
+    private final String name;
 
     public Station() {
+        this(EMPTY_ID, EMPTY_NAME);
```
기본생성자가 비즈니스적으로 사용되지않는다면 private으로 막아버려도 좋을 것 같습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/SpringLineService.java:
```diff
@@ -16,27 +17,32 @@ public SpringLineService(LineDao lineDao) {
     }
 
     @Override
+    @Transactional(readOnly = true)
```
클래스 레벨에 @Transactional을 설정할 수 있는데요. 모든 메서드에 적용됩니다.
클래스레벨을 `@Transactional(readOnly = true)`에 걸어놓고, 변경이 있을곳에 `@Transactional`를 걸어보는건 어떨까요?

트랜잭션 적용 우선순위가 메소드, 클래스 순이라서 코드작성하기가 수월할 것 같아요.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Station.java:
```diff
@@ -23,5 +34,29 @@ public Long getId() {
         return name;
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof Station)) {
+            return false;
+        }
+        Station station = (Station) o;
+        return id.equals(station.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id);
+    }
```
👍

hongbin-dev commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,160 @@
+        final ExtractableResponse<Response> createResponse2 = createLine("2호선", "bg-green-600");
+
+        return Stream.of(
+                DynamicTest.dynamicTest("생성된 노선 목록을 불러온다", () -> {
+                    final ExtractableResponse<Response> response = getLines();
+
+                    assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+                }),
+                DynamicTest.dynamicTest("생성된 노선이 저장한 노선과 일치한지 확인한다.", () -> {
+                    final ExtractableResponse<Response> response = getLines();
+                    final List<Long> expectedLineIds = Stream.of(createResponse1, createResponse2)
+                            .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
+                            .collect(Collectors.toList());
+                    final List<Long> resultLineIds = response.jsonPath().getList(".", LineResponse.class).stream()
+                            .map(it -> it.getId())
+                            .collect(Collectors.toList());
+
+                    assertThat(resultLineIds).containsAll(expectedLineIds);
+                })
```
마이너하지만 

```suggestion
                DynamicTest.dynamicTest("생성된 노선 목록을 불러온다", () -> {
                    final ExtractableResponse<Response> response = getLines();

                    assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
                }),
                
                DynamicTest.dynamicTest("생성된 노선이 저장한 노선과 일치한지 확인한다.", () -> {
                    final ExtractableResponse<Response> response = getLines();
                    final List<Long> expectedLineIds = Stream.of(createResponse1, createResponse2)
                            .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
                            .collect(Collectors.toList());
                    final List<Long> resultLineIds = response.jsonPath().getList(".", LineResponse.class).stream()
                            .map(it -> it.getId())
                            .collect(Collectors.toList());

                    assertThat(resultLineIds).containsAll(expectedLineIds);
                })
```

중간에 개행을 넣어주면 가독성이 좋아질 것 같네요

hongbin-dev reviewed (approved):
안녕하세요 스컬!
리뷰가 늦어졌네요 🥲 피드백 반영 잘 해주셨습니다. 특히 다이나믹테스트 적용하니 가독성이 좋아졌네요.
머지해도 좋을 것 같아요~ 다음단계에서 만나요
